src/algebra/module/submodule/basic.lean

```diff
@@ -196,7 +196,7 @@ instance : has_zero p := ⟨⟨0, zero_mem _⟩⟩
 instance : inhabited p := ⟨0⟩
 instance [has_smul S R] [has_smul S M] [is_scalar_tower S R M] :
   has_smul S p := ⟨λ c x, ⟨c • x.1, smul_of_tower_mem _ c x.2⟩⟩
-
+instance {V : Type*} [has_vadd M V] : has_vadd p V := ⟨ λ s v, (s : M) +ᵥ v ⟩
```

Comment:
Ideally we'd repeat the additive version of

https://github.com/leanprover-community/mathlib/blob/22131150f88a2d125713ffa0f4693e3355b1eb49/src/group_theory/submonoid/operations.lean#L1080-L1144

including the header comment. Note that we already have all of these instances on `add_submonoid`, they just haven't been transferred to `submodule`.

Reply:
This is what I come out with 
```lean
import algebra.module.submodule.basic

/-! ### Actions by `submodule`s

These instances tranfer the action by an element `m : M` of a module `M` written as `m +ᵥ a` onto
the action by an element `s : S` of a submodule `S : submodule M` such that `s +ᵥ a = (s : M) +ᵥ a`.

These instances work particularly well in conjunction with `module.to_add_action`, enabling
`s +ᵥ m` as an alias for `↑s + m`.
-/

namespace submodule
variables {M R : Type*} {α β : Type*}

variables [add_comm_monoid M] [semiring R] [module R M]

instance [has_vadd M α] (S : submodule R M) : has_vadd S α := has_vadd.comp _ S.subtype

instance vadd_comm_class
  [has_vadd M β] [has_vadd α β] [vadd_comm_class M α β] (S : submodule R M) :
  vadd_comm_class S α β := ⟨λ a, (vadd_comm (a : M) : _)⟩

lemma vadd_def [has_vadd M α] {S : submodule R M} (g : S) (m : α) : g +ᵥ m = (g : M) +ᵥ m := rfl

instance [has_vadd M α] [has_faithful_vadd M α] (S : submodule R M) :
  has_faithful_vadd S α := ⟨λ x y h, subtype.ext $ eq_of_vadd_eq_vadd h⟩

/-- The action by a submodule is the action by the underlying submodule. -/
instance [add_action M α] (S : submodule R M) : add_action S α :=
  add_action.comp_hom _ S.subtype.to_add_monoid_hom

end submodule
```

Reply:
Is that what you had in mind?

Reply:
Yes, that looks quite reasonable. I think you should be able to write a lemma about `vsub` as well, although I guess we're missing that for `add_subgroup` too. These might be good in their own PR.

Reply:
I added the code above to `algebra.module.submodule.basic`. I think indeed that the part about `vsub` will have to wait for its own PR later on since it involves a lot more work.